src/orange/OrangeParams.cc

```diff
@@ -134,66 +134,89 @@ OrangeParams::OrangeParams(OrangeInput input)
-            = host_data.unit_indexer_data.volumes[AllVals{}].back();
-
-        auto surfs = std::visit(get_num_surfaces, u);
-        auto vols = std::visit(get_num_volumes, u);
-
-        ui_surf.push_back(surface_offset + surfs);
-        ui_vol.push_back(volume_offset + vols);
+        auto ui_surf = make_builder(&host_data.universe_indexer_data.surfaces);
+        auto ui_vol = make_builder(&host_data.universe_indexer_data.volumes);
+        ui_surf.push_back(0);
+        ui_vol.push_back(0);
+
+        auto get_num_surfaces = Overload{
+            [](UnitInput const& u) -> size_type { return u.surfaces.size(); },
+            [](RectArrayInput const& r) -> size_type {
+                return r.daughters.size();
+            }};
+
+        auto get_num_volumes = Overload{
+            [](UnitInput const& u) -> size_type { return u.volumes.size(); },
+            [](RectArrayInput const& r) -> size_type {
+                return std::accumulate(r.grid.begin(),
+                                       r.grid.end(),
+                                       size_type(0),
+                                       [](size_type acc, const auto& vec) {
+                                           return acc + vec.size();
+                                       });
+            }};
+
+        for (auto const& u : input.universes)
+        {
+            using AllVals = AllItems<size_type, MemSpace::native>;
+
+            auto surface_offset
+                = host_data.universe_indexer_data.surfaces[AllVals{}].back();
+            auto volume_offset
+                = host_data.universe_indexer_data.volumes[AllVals{}].back();
+
+            auto surfs = std::visit(get_num_surfaces, u);
+            auto vols = std::visit(get_num_volumes, u);
+
+            ui_surf.push_back(surface_offset + surfs);
+            ui_vol.push_back(volume_offset + vols);
+        }
     }
 
-    detail::UnitInserter insert_unit(&host_data);
-    detail::RectArrayInserter insert_rect_array(&host_data);
-
-    auto insert_universe = Overload{
-        [&insert_unit](UnitInput const& u) {
-            CELER_VALIDATE(u,
-                           << "simple unit '" << u.label
-                           << "' is not properly constructed");
-
-            insert_unit(u);
-        },
-        [&insert_rect_array](RectArrayInput const& r) {
-            CELER_VALIDATE(r,
-                           << "rect array '" << r.label
-                           << "' is not properly constructed");
-
-            insert_rect_array(r);
-        },
-    };
+    // Create universe_types and universe_indices vectors
+    {
+        auto u_types_builder = make_builder(&host_data.universe_types);
+        auto u_indices_builder = make_builder(&host_data.universe_indices);
+
+        std::vector<size_type> current_indices(
+            static_cast<size_t>(UniverseType::size_), 0);
+
+        for (auto const& u : input.universes)
+        {
+            auto u_type_idx = u.index();
```

Comment:
This is dangerous because it relies implicitly on the `variant` ordering matching the `UniverseType` ordering. I think my WIP refactor should make this a little more elegant though.

Reply:
This was the only way I could think to do this, but I will be interested to see what you came up with.